### DISCUSSION TITLE
type vs interface

### DISCUSSION THREAD

**Discussion by michoball** [wanted-9th-3team/frontend-study#33] [프론트 기술 질문 공간]
typescript 에서 type  과 interface의 차이점에 대해 알려주세요

**Comment by skyhanull** [comment]
type과 interface의 차이란?
type과 interface 모두 typescript에서 타입을 지정해줄 때 사용합니다
이 두가지는 거의 비슷하게 사용되지만 약간의 차이점이 존재합니다.
1. 확장방법
interface는 타입을 확장할 때 extends를 써서 확장을 합니다.
type는 타입을 확장할 때 &를 써서 확장을 합니다.

2. 선언적확장
interface에서 할 수 있는 대부분의 기능들은 type에서 가능하지만, 한 가지 중요한 차이점은 type은 새로운 속성을 추가하기 위해서 다시 같은 이름으로 선언할 수 없지만, interface는 항상 선언적 확장이 가능하다는 것이다.

3. computed property name사용유무
interface는 객체로 사용이 가능하지만 computed property name은 사용이 불가능합니다
하지만 type은 computed property name사용이 가능합니다
computed property name은 인덱스 시그니쳐와 다른것입니다!

**Comment by bona373737** [comment]
type alias와 interface는 둘 다 타입에 이름을 부여해주는 기능이지만 사용상에 몇가지 차이점이 있다.

차이점1. 이름부여가 가능한 타입의 대상
  type alias : 모든(any)타입에 이름을 부여가 가능하다. 
  interface : 오직 객체타입에만 가능하다. 

차이점2. 타입확장방식
  type alias : &으로 확장
  interface : extends로 확장

차이점2. 동일한 이름으로 확장가능 여부
  type alias : 동일한 이름으로 type선언불가
  interface : 동일한 이름으로 interface 선언 가능

TypeScript 팀에서는 확장이 용이하다는 이유로 interface의 사용을 권장하고 있다. 
type과 interface간의 큰 성능의 차이는 없다.

**Comment by michoball** [comment]
type alias vs interface

공식문서에는 다음과 같이 나와있다. 

Type aliases and interfaces are very similar, and in many cases you can choose between them freely. Almost all features of an interface are available in type, the key distinction is that a type cannot be re-opened to add new properties vs an interface which is always extendable.

둘은 매우 흡사하고 대부분의 경우 개발자의 취향에 따라 선택적으로 사용하면 된다. 

단 차이점이 있다면 type alias는 모든 타입에 사용이 가능한 반면, 새로운 속성을 추가하는데 닫혀있고 interface는 객체 타입에만 사용할 수 있지만 속성을 추가하는데 열려있다는 것이다. 

interface의 경우 extends를 사용하거나 다시 interface로 동일한 타입에 속성을 넣어서 확장성있게 사용할 수 있다. 

```typescript
interface Point {
  x: number;
  y: number;
}
 
interface Point {
  z: number;
}
 
하면 

const direction : Point = {
  x: 1;
  y: 10;
  z:  -30;
}; 
해도 됨 ( 선언 병합이 가능하다. )
```

하지만 type alias로는 재 정의하여 속성을 추가할 수 없다. 추가하기 위해서는 & 를 사용해 새로운 type을 만들어야한다.

```typescript
type Point = {
  x: number;
  y: number;
}
하면
type NewPoint  = Point & {
  z: number;
}

해서 NewPoint를 사용해야함
```
위 선언 병합을 할 수 있다는 interface의 특징 때문에 interface가 좀 더 확장성있게 사용가능하다. 그래서 대부분의 경우 잘모르겠으면 interface를 사용하는것이 좋다고 생각한다.

**Comment by WooGie911** [comment]
타입과 인터페이스는 대부분의 경우에는 차이점이 없습니다.

인터페이스의 가장 큰 특징은 선언 병합을 통해 속성을 확장시킬 수 있다는 점입니다.

다만 복잡한 타입을 다룰 때는 type을 사용하는 것이 좋습니다.

인터페이스를 사용해야 하는 경우를 간단하게 정리하면
- 선언 병합(자동 병합)의 이점을 활용해야 하는 경우
- 객체 타입을 정의하거나, 타입을 사용할 필요가 없을 경우

타입을 사용해야 하는 경우는
- 원시 타입을 정의할 경우
- 튜플 타입을 정의할 경우
- 함수 타입을 정의할 경우
- 유니온 타입을 정의할 경우
- 매핑된 타입을 정의할 경우

로 나눌 수 있습니다.

**Comment by BaekYuri** [comment]
타입과 인터페이스는 타입을 부여해주는 문법입니다.
다른 점은, 확장을 할 때, 인터페이스는 extends를 사용하고 타입은 &을 사용합니다.
또한 인터페이스는 선언전 확장을 할 수 있지만, 타입에서 선언적 확장을 하려고 하면 에러가 납니다.
또한 인터페이스는 객체에만 타입을 선언할 수 있지만, 타입은 모든 타입에 타입을 선언할 수 있습니다.
